### PR TITLE
Add Gaussian noise to Odometry Publisher

### PR DESCRIPTION
**Pull request by adityapande-1995**
🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

# 🎉 New feature

## Summary
This PR adds Gaussian noise to the published odometry message, and a new publisher which publishes ``OdometryWithCovariance`` messages. 

Depends on https://github.com/ignitionrobotics/ign-msgs/pull/224.
This PR in ``ros_ign`` makes it compatible with ``geometry_msgs::odometry`` : https://github.com/ignitionrobotics/ros_ign/pull/222

## Test it
Added a test case that checks the noise statistically and the diagonal elements of the covariance matrix.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

**Comment by adityapande-1995** [comment]
## Concerns to be addressed before moving ahead: 
1. This PR adds a second publisher for ``OdometryWithCovariance`` along with the existing one, to maintain API. Is this okay or should this PR be based to ``ign-gazebo7`` ?
2. This is trying to replicate the behavior in ``gazebo_ros_p3d`` plugin. Originally, the noise was added only to the ``twist`` and NOT the ``pose``, yet the covariance matrix is populated with diagonal elements for both twist and pose. Should it be a zero matrix for pose ? I've replicated the same behavior here.

File: src/systems/odometry_publisher/OdometryPublisher.cc [hunk 1]
```diff
@@ -158,6 +165,11 @@ void OdometryPublisher::Configure(const Entity &_entity,
         "rpy_offset"));
   }
 
+  if (_sdf->HasElement("gaussian_noise"))
+  {
+    this->dataPtr->gaussianNoise = _sdf->Get<double>("gaussian_noise");
```
Comment (jacobperron) [inline src/systems/odometry_publisher/OdometryPublisher.cc:170]: Should the new sdf elements (`gaussian_noise` and `odom_covariance_topic`) be documented somewhere?

File: src/systems/odometry_publisher/OdometryPublisher.cc [hunk 2]
```diff
@@ -95,6 +99,9 @@ class ignition::gazebo::systems::OdometryPublisherPrivate
 
   /// \brief Allow specifying constant xyz and rpy offsets
   public: ignition::math::Pose3d offset = {0, 0, 0, 0, 0, 0};
+
+  /// \brief Gaussian noise
+  public: double gaussianNoise = 0;
```
[the remark this answers is not in the document]
Comment (adityapande-1995) [reply src/systems/odometry_publisher/OdometryPublisher.cc]: Done

File: src/systems/odometry_publisher/OdometryPublisher.cc (hunk 1 above)
Comment (adityapande-1995) [reply src/systems/odometry_publisher/OdometryPublisher.cc:170]: Done

File: src/systems/odometry_publisher/OdometryPublisher.cc [hunk 3]
```diff
@@ -210,6 +228,16 @@ void OdometryPublisher::Configure(const Entity &_entity,
   }
   this->dataPtr->odomPub = this->dataPtr->node.Advertise<msgs::Odometry>(
       odomTopicValid);
+
+  std::string odomCovTopicValid {transport::TopicUtils::AsValidTopic(odomCovTopic)};
+  if (odomCovTopicValid.empty())
+  {
+    ignerr << "Failed to generate odom topic ["
+           << odomCovTopic << "]" << std::endl;
+    return;
```
[the remark this answers is not in the document]
Comment (adityapande-1995) [reply src/systems/odometry_publisher/OdometryPublisher.cc]: Done

File: src/systems/odometry_publisher/OdometryPublisher.cc [hunk 4]
```diff
@@ -374,6 +417,63 @@ void OdometryPublisherPrivate::UpdateOdometry(
...
+        mutable_covariance()->add_data(gn2);
+    }
+    else {
+      msg_covariance.mutable_pose_with_covariance()->
+        mutable_covariance()->add_data(0);
+      msg_covariance.mutable_twist_with_covariance()->
+        mutable_covariance()->add_data(0);
+    }
+  }
+  this->odomCovPub.Publish(msg_covariance);
```
[the remark this answers is not in the document]
Comment (adityapande-1995) [reply src/systems/odometry_publisher/OdometryPublisher.cc]: Added a check to see if the publisher is valid.

**Comment by adityapande-1995** [comment]
CI is unable to find the ``ign-msgs`` headers for the new messages ``odometry_with_covariance``

**Comment by nuclearsandwich** [comment]
@osrf-jenkins retest this please

**Review by jacobperron — APPROVED** [review]

**Review by chapulina — APPROVED** [review]
LGTM! I took the liberty of making a few commits directly to this branch to get it through CI and into the upcoming release:

* 1d2033f79b1cac3e9d60740bdfa77d133e80d05e: Require `ign-msgs 8.3`
* cf78741735aa8b240ce22344ece94866eb2e2f00 and 17e77b0d69c87bb1571334bbfc313a2479008e5a: style fixes that aren't caught by our linters (it may be worth it taking a look at them to keep this in mind for future PRs)

**Comment by adityapande-1995** [comment]
The test failing in focal is a flaky one.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-releases-2022-04-27-fortress-citadel/1389/1

**Pull request merged**